jdaviz/app.py

```diff
@@ -1341,51 +1347,87 @@ def remove(stack_items):
+            [data] = [x for x in self.data_collection if x.label == data_label]
+            color_cycler = ColorCycler()
+            viewer.add_data(data, percentile=95, color=color_cycler())
+
+            add_data_message = AddDataMessage(data, viewer,
+                                              viewer_id=viewer_id,
+                                              sender=self)
+            self.hub.broadcast(add_data_message)
+
+        # set visibility state of all applicable layers
+        for layer in viewer.layers:
+            if layer.layer.data.label == data_label:
+                if visible and not layer.visible:
+                    layer.visible = True
+                    layer.update()
+                else:
+                    layer.visible = visible
 
-    def vue_data_item_visibility(self, event):
-        viewer_id, item_id, visible = event['id'], event['item_id'], event['visible']
-        viewer_item = self._viewer_item_by_id(viewer_id)
+        # if replace, do another loop (we do a second loop to ensure the visible layer is added
+        # first BEFORE other layers are removed)
+        if replace:
+            for layer in viewer.layers:
+                if layer.layer.data.label != data_label:
+                    layer.visible = False
 
-        replace = event.get('replace', False)
+        # update data menu - selected_data_items should be READ ONLY, not modified by the user/UI
         selected_items = viewer_item['selected_data_items']
+        data_id = self._data_id_from_label(data_label)
+        selected_items[data_id] = 'visible' if visible else 'hidden'
+        if replace:
+            for id in selected_items:
+                if id != data_id:
+                    selected_items[id] = 'hidden'
 
-        if replace and visible:
-            # trace items remain selected, even if the mode is in replace
-            keep_selected = [k for k, v in viewer_item['selected_data_items'].items()
-                             if v != 'hidden' and
-                             self._get_data_item_by_id(k).get('type') == 'trace']
-            selected_items = {id: 'visible' if id == item_id or id in keep_selected else 'hidden' for id in selected_items.keys()}  # noqa
-        else:
-            selected_items[item_id] = 'visible' if visible else 'hidden'
+        # Sets the plot axes labels to be the units of the most recently
+        # active data.
+        if len(viewer_data_labels) > 0 and self._jdaviz_helper._in_batch_load == 0:
+            active_data = self.data_collection[viewer_data_labels[0]]
+            if (hasattr(active_data, "_preferred_translation")
+                    and active_data._preferred_translation is not None):
+                self._set_plot_axes_labels(viewer_id)
 
-        self._update_selected_data_items(viewer_id, selected_items)
+            if self.config == 'imviz':
+                viewer.on_limits_change()  # Trigger compass redraw
```

Comment:
I had to modify this logic slightly to get it to work in the new method.  Is the assumption of using the first data layer label valid here?  It seems this logic was originally implemented in #191.

Apparently not because it seems to break the axes labeling in Specviz... I'll investigate and fix that before marking this for review.